realm/realm-library/src/objectServer/java/io/realm/SyncSession.java

```diff
@@ -387,7 +388,12 @@ private void waitForChanges(int direction) throws InterruptedException {
 
                 throw new ObjectServerError(ErrorCode.UNKNOWN, errorMsg + " Has the SyncClient been started?");
             }
-            wrapper.waitForServerChanges();
+            try {
+                wrapper.waitForServerChanges();
+            } catch(InterruptedException e) {
+                waitingForServerChanges.set(null); // Ignore any results being sent if the wait was interrupted.
```

Comment:
This is not enough.
There is a chance that users add another listener after the cancellation.

```
SyncSession session = ...;
try {
    session.downloadAllServerChanges();
} catch (InterruptedException e) {
    // cancelled
}

...  // so some works
session.uploadAllLocalChanges(); // this may return too early because of callback caused by downloadAllServerChanges() before
```